### ISSUE TITLE
Possible GPU Memory Leak - Mem Usage Keeps on Increasing with Every Iteration

### ISSUE THREAD

**Issue by MXGray** [NVIDIA/tacotron2#39]
I'm on Ubuntu 16.04, CUDA 9.1, CUDNN 7.1, one P6000 GPU (compute 6.1) with 16 Gb VRAM, Python 3.6, Pytorch 0.4 and all other required libraries are successfully installed and properly working.
Tried batch sizes 4, 8, 12, 16, 24, 44 and 48.
Always get out of memory error after some time.

Training starts, and GPU VRAM usage keeps on increasing with every iteration.
I tried correcting a couple of lines in layers.py and another line in train.py following what was said in deprecation warnings.
Still get out of memory errors.

Tried disabling CUDNN in hparams after reading something about memory leaks with Pytorch and CUDNN.
Still get out of memory errors.

I have a private dataset that mimicks the exact attributes, properties and structure of LJSpeech.
Tried using all 480K WAV train files, 20K WAV test files and 5K WAV validation files.
Also tried using 12.5K WAV training files, 1K WAV test files and 500 WAV validation files (same as LJSpeech).
Still get out of memory errors.

What could be the problem?
Would greatly appreciate some help regarding this.
Thanks in advance!

**Comment by rafaelvalle** [comment]
Thank you for reporting this. 
Can you build a docker container using this [docker file](https://github.com/NVIDIA/tacotron2/blob/master/Dockerfile) and let us know if the error persists?

Regarding running with CUDA 9.1 and CUDNN 7.1:
Are you using the default values in hparams?
Can you pin point where the increase in memory is occurring with the default hparams? 
My two guesses are that it occurs at some `+=` operation or that it occurs in some operation in the  LocationLayer in model.py.

**Comment by MXGray** [comment]
@rafaelvalle
This happened again:

Train loss 254 0.974303 Grad Norm 1.096113 18.54s/it
THCudaCheck FAIL file=pytorch_1524584710464/work/aten/src/THC/generic/THCStorage.cu line=58 error=2 : out of memory
Traceback (most recent call last):
  File "train.py", line 291, in <module>
    args.warm_start, args.n_gpus, args.rank, args.group_name, hparams)
File "train.py", line 216, in train
    y_pred = model(x)
  File "torch/nn/modules/module.py", line 491, in __call__ result = self.forward(*input, **kwargs)
  File "model.py", line 510, in forward encoder_outputs, targets, memory_lengths=input_lengths)
File "python3.6/site-packages/torch/nn/modules/module.py", line 491, in __call__ result = self.forward(*input, **kwargs)
File "model.py", line 403, in forward decoder_input)
File "model.py", line 363, in decode attention_weights_cat, self.mask)
  File "python3.6/site-packages/torch/nn/modules/module.py", line 491, in __call__ result = self.forward(*input, **kwargs)
  File "model.py", line 77, in forward attention_hidden_state, processed_memory, attention_weights_cat)
File "model.py", line 60, in get_alignment_energies processed_query + processed_attention_weights + processed_memory))
RuntimeError: cuda runtime error (2) : out of memory at pytorch_1524584710464/work/aten/src/THC/generic/THCStorage.cu:58 (AIWN9)

Will now try the updated dockerfile ...

**Comment by MXGray** [comment]
@rafaelvalle
Still got the error even with the dockerfile. Here's what happened with batch_size 12:

Train loss 107 0.993192 Grad Norm 0.723081 11.45s/it
THCudaCheck FAIL file=/pytorch/aten/src/THC/generic/THCStorage.cu
line=58 error=2 : out of memory
Exception ignored in: <bound method _DataLoaderIter.__del__ of <torch.utils.data.dataloader._DataLoaderIter object at 0x7f4070db6da0>>
Traceback (most recent call last):
  File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 349, in __del__ self._shutdown_workers()
File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 328, in _shutdown_worker self.worker_result_queue.get()
File "/home/MX9/anaconda3/lib/python3.6/multiprocessing/queues.py", line 337, in get return _ForkingPickler.loads(res)
  File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/multiprocessing/reductions.py", line 70, in rebuild_storage_fd fd = df.detach()
  File "/home/MX9/anaconda3/lib/python3.6/multiprocessing/resource_sharer.py", line 57, in detach with _resource_sharer.get_connection(self._id) as conn:
  File "/home/MX9/anaconda3/lib/python3.6/multiprocessing/resource_sharer.py", line 87, in get_connection c = Client(address, authkey=process.current_process().authkey)
  File "/home/MX9/anaconda3/lib/python3.6/multiprocessing/connection.py", line 494, in Client deliver_challenge(c, authkey)
  File "/home/MX9/anaconda3/lib/python3.6/multiprocessing/connection.py", line 722, in deliver_challenge response = connection.recv_bytes(256)        # reject large message
File "/home/MX9/anaconda3/lib/python3.6/multiprocessing/connection.py",line 216, in recv_bytes buf = self._recv_bytes(maxlength)
  File "/home/MX9/anaconda3/lib/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes buf = self._recv(4)
  File "/home/MX9/anaconda3/lib/python3.6/multiprocessing/connection.py", line 379, in _recv chunk = read(handle, remaining) ConnectionResetError: [Errno 104] Connection reset by peer
Traceback (most recent call last):
  File "train.py", line 291, in <module> args.warm_start, args.n_gpus, args.rank, args.group_name, hparams)
  File "train.py", line 216, in train
    y_pred = model(x)
  File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 491, in __call__ result = self.forward(*input,**kwargs)
  File "/opt/AITTSv3/nv-wavenet/pytorch/tacotron2/model.py", line 512, in forward mel_outputs_postnet = self.postnet(mel_outputs) 
File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py",line 491, in __call__ result = self.forward(*input, **kwargs)
File "/opt/AITTSv3/nv-wavenet/pytorch/tacotron2/model.py", line 144, in forward x = self.dropout(F.tanh(self.convolutions[i](x)))
  File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 491, in __call__ result = self.forward(*input, **kwargs) 
File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/nn/modules/dropout.py", line 53,in forward return F.dropout(input, self.p, self.training, self.inplace) 
File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/nn/functional.py",line 552, in dropout return _functions.dropout.Dropout.apply(input, p, training, inplace)
  File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/nn/_functions/dropout.py", line 36, in forward ctx.noise = cls._make_noise(input)
File "/home/MX9/anaconda3/lib/python3.6/site-packages/torch/nn/_functions/dropout.py", line 10, in _make_noise return input.new().resize_as_(input)
RuntimeError: cuda runtime error (2) : out of memory at /pytorch/aten/src/THC/generic/THCStorage.cu:58

Perhaps max length? Would greatly appreciate some help. :)

**Comment by MXGray** [comment]
Still haven't found the cause of this.
I've tested two machines already.
Machine A has 1 P6000 16Gb VRAM.
Machine 2 has 1 Tesla V100 16Gb VRAM.
...
Have also placed torch.cuda.empty_cache() and gc.collect() in main loop.
Still get CUDA out of memory error.
Have tested downgrading to CUDA 8 and CUDNN 6, installed Pytorch 0.4.0 CUDA80 and Tensorflow 1.3 (and 1.6) for machine A.
CUDA out of memory error still persists.
...
Anybody encountering the same problem with a private dataset?

**Comment by MXGray** [comment]
Finally! I've identified the cause of this problem.
Despite running my getlength.py script through my raw dataset to separate a dev dataset, it still got mixed up with some audio files with lengths beyond the largest lengths of the files in the LJSpeech-1.1 dataset.
Now running 64 batch size in machine with 1 Tesla V100 16Gb VRAM GPU.  :)
Closing this now ...
BTW, largest length of files in LJSpeech-1.1 is 10092 ...
